cmd/ecc/buffer.go

```
@@ -0,0 +1,167 @@
+func NewBuffer(shards [][]byte, parity int) *Buffer {
+	if len(shards) > 0 {
+		for _, shard := range shards {
+			if len(shard) != cap(shard) {
+				panic("ecc: Buffer requires that each shard has equal len and cap")
+			}
+			if len(shards[0]) != len(shard) {
+				panic("ecc: Buffer requires that each shard has the same len")
+			}
+		}
+	}
+	return &Buffer{
+		shards: shards,
+		data:   shards[:len(shards)-parity],
+		parity: shards[len(shards)-parity:],
+	}
+}
+
+// IsDataMissing returns true if at least one
+// data shard is marked as missing. A data shard
+// is considered as not present if its length
+// len(shard) == 0.
+func (b *Buffer) IsDataMissing() bool {
+	dataShards := len(b.shards) - len(b.parity)
+	for _, shard := range b.shards[:dataShards] {
+		if len(shard) == 0 {
+			return true
+		}
+	}
+	return false
+}
+
+// Empty marks the buffer as empty such that
+// IsEmpty() returns true.
+func (b *Buffer) Empty() *Buffer {
+	b.offset = 0
+	b.data = b.data[:0]
+	return b
+}
+
+// IsEmpty returns true if the buffer does not
+// hold any actual data. In particular, copying
+// an empty buffer to a slice (using CopyTo)
+// copies no data.
+func (b *Buffer) IsEmpty() bool { return b.offset == 0 && len(b.data) == 0 }
+
+// Reset resets the buffer to its initial state.
+// In particular, a reseted buffer is not empty.
+func (b *Buffer) Reset() *Buffer {
+	for i := range b.shards {
+		b.shards[i] = b.shards[i][:cap(b.shards[i])]
+	}
+	b.data = b.shards[:len(b.shards)-len(b.parity)]
+	b.parity = b.shards[len(b.data):]
+
+	b.buffer = nil
+	b.offset = 0
+	return b
+}
+
+// Skip skips the next n (actual data) bytes hold
+// by the buffer. If n is greater than the number
+// of remaining data bytes hold by the buffer, Skip
+// will only skip as many (actual data) bytes as
+// available. In particular, trying to skip n > 0
+// bytes on an empty buffer is a NOP.
+func (b *Buffer) Skip(n int) *Buffer {
```

I don't like the design of this. If the caller asks to skip more bytes than available, it indicates an error in the code and silently ignoring will just make the caller think everything is ok.

```suggestion
func (b *Buffer) Skip(n int) error {
```
Since you are modifying the receiver there is no point in returning it as well.

---

1. DSL Style - as done by `bytes.Buffer`
2. IMO that is not correct. Skipping more bytes is completely fine - the buffer is just empty afterwards and `Skip` returns how many bytes where skipped. Otherwise, the caller needs to know in advance how many bytes are in the buffer.

---

1. [bytes.Buffer](https://golang.org/pkg/bytes/#Buffer) does not have DSL style. I'm not sure what you mean? Nothing in the stdlib has AFAIK.

2. It indicates an error by the caller. If they want to check, adding a `Len() int` method should be trivial. This is one of the places where returning `io.ErrUnexpectedEOF` makes sense.

Silently swallowing errors is dangerous and can lead to silent data corruption.